File: docs/perf-comparison/sieve.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <time.h>

int64_t sieve(int64_t limit) {
    int64_t word_cnt = (limit + 64) / 64;
    uint64_t* bits = calloc(word_cnt, sizeof(uint64_t));
    if (!bits) return -1;

    for (int64_t i = 2; i * i <= limit; i++) {
        if (!((bits[i >> 6] >> (i & 63)) & 1)) {
            for (int64_t j = i * i; j <= limit; j += i)
                bits[j >> 6] |= (1UL << (j & 63));
        }
    }

    int64_t count = 0;
    for (int64_t i = 2; i <= limit; i++) {
        if (!((bits[i >> 6] >> (i & 63)) & 1))
            count++;
    }
    free(bits);
    return count;
}

int main(int argc, char** argv) {
    int64_t limit = argc > 1 ? atol(argv[1]) : 1000000;
    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);
    int64_t result = sieve(limit);
    clock_gettime(CLOCK_MONOTONIC, &end);
    double ms = (end.tv_sec - start.tv_sec) * 1000.0
              + (end.tv_nsec - start.tv_nsec) / 1000000.0;
    printf("C,%lld,%lld,%.1f\n", (long long)limit, (long long)result, ms);
    return 0;
}

File: docs/perf-comparison/sieve.py
import sys
import time

def sieve(limit):
    word_cnt = (limit + 64) // 64
    bits = bytearray(word_cnt * 8)

    i = 2
    while i * i <= limit:
        wi = i >> 6
        bi = i & 63
        word = int.from_bytes(bits[wi*8:(wi+1)*8], 'little')
        if (word >> bi) & 1 == 0:
            j = i * i
            while j <= limit:
                wj = j >> 6
                bj = j & 63
                off = wj * 8
                w = int.from_bytes(bits[off:off+8], 'little')
                w |= 1 << bj
                bits[off:off+8] = w.to_bytes(8, 'little')
                j += i
        i += 1

    count = 0
    for i in range(2, limit + 1):
        wi = i >> 6
        bi = i & 63
        off = wi * 8
        w = int.from_bytes(bits[off:off+8], 'little')
        if (w >> bi) & 1 == 0:
            count += 1
    return count

if __name__ == '__main__':
    limit = int(sys.argv[1]) if len(sys.argv) > 1 else 1000000
    start = time.perf_counter()
    result = sieve(limit)
    us = (time.perf_counter() - start) * 1_000_000
    print(f"Python,{limit},{result},{us:.0f}")
